File: leetcode/208-implement-trie-prefix-tree.py
# https://leetcode.com/problems/implement-trie-prefix-tree/ (top-interview-questions)

class Node:
    def __init__(self):
        self.childs = dict()
        self.is_word_end = False

class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        curr = self.root
        for char in word:
            if char not in curr.childs:
                curr.childs[char] = Node()
            curr = curr.childs[char]
        curr.is_word_end = True

    def search(self, word: str) -> bool:
        curr = self.root
        for char in word:
            if char not in curr.childs:
                return False
            curr = curr.childs[char]
        return curr.is_word_end

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for char in prefix:
            if char not in curr.childs:
                return False
            curr = curr.childs[char]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: leetcode/208-implement-trie-prefix-tree.js
// https://leetcode.com/problems/implement-trie-prefix-tree/ (top-interview-questions)

class Node {
  constructor() {
    this.childs = {};
    this.isWordEnd = false;
  }
}

var Trie = function() {
  this.root = new Node();
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
  
  let node = this.root;
  for (const char of word) {
    if (!node.childs.hasOwnProperty(char)) {
      node.childs[char] = new Node();
    }
    node = node.childs[char];
  }

  node.isWordEnd = true;

};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
  
  let node = this.root;
  for (const char of word) {
    if (!node.childs.hasOwnProperty(char)) {
      return false;
    }
    node = node.childs[char];
  }

  return node.isWordEnd;
};

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
  let node = this.root;
  for (const char of prefix) {
    if (!node.childs.hasOwnProperty(char)) {
      return false;
    }
    node = node.childs[char];
  }

  return true;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */
